---
title: "R Notebook"
output: html_notebook
---


Make lists of relevant genes for each contrast to test PaintOmics 


```{r}
wd <- getwd() #"C:\\Users\\Mukta Palshikar\\Box\\multiomics integration manuscript\\data\\paintomics\\"
setwd(wd)
wd <- paste0(wd, "/")
print(wd)
```

# convert to tab-separated files 

```{r}
proteomics = read.csv("bonita_proteomics.csv")
colnames(proteomics)[1] = "Gene"
write.table(proteomics, "paintomics_proteomics.txt", sep="\t", row.names=FALSE)

phosphoproteomics = read.csv("bonita_phosphoproteomics.csv")
colnames(phosphoproteomics)[1] = "Gene"
write.table(phosphoproteomics, "paintomics_phosphoproteomics.txt", sep="\t", row.names=FALSE)

transcriptomics = read.csv("bonita_transcriptomics.csv")
colnames(transcriptomics)[1] = "Gene"
write.table(transcriptomics, "paintomics_transcriptomics.txt", sep="\t", row.names=FALSE)
```

# relevant genes - de genes from each contrast. used ReactomeGSA to perform DE analysis with limma/eBayes


See the sheet 'Explanation' in DE_analysis_20221003.xlsx

```{r}
require(readxl)

explanation <- read_excel("C:\\Users\\Mukta\ Palshikar\\Documents\\GitHub\\moBONITA\\data\\paintomics\\DE_analysis_20221003.xlsx", sheet = 'Explanation')

for (i in 1:nrow(explanation)){
  #print(explanation$Name[i])
  temp = read_excel("C:\\Users\\Mukta\ Palshikar\\Documents\\GitHub\\moBONITA\\data\\paintomics\\DE_analysis_20221003.xlsx", sheet = explanation$Name[i])
  fileName = paste0(explanation$Dataset[i], "_", explanation$Contrast[i], ".csv")
  print(fileName)
  write.csv(temp, fileName)
  fileName = paste0("Relevant_", fileName)
  temp2 = as.data.frame(temp$Identifier[temp$logFC>quantile(temp$logFC)[4]])
  colnames(temp2) = "Gene"
  write.csv(temp2, fileName, row.names=FALSE)
}

```

# old code, do not run

```{r, eval=FALSE}

makePaintOmics <- function(contrast){
  # proteomics
  proteomics <- read.csv(paste(wd, "proteomics/", contrast, sep = ""))
  colnames(proteomics) <- c('Gene','proteomics')
  # phosphoproteomics
  phosphoproteomics <- read.csv(paste(wd, "phosphoproteomics/", contrast, sep = ""))
  colnames(phosphoproteomics) <- c('Gene','phosphoproteomics')
  # transcriptomics
  transcriptomics <- read.csv(paste(wd, "transcriptomics/", contrast, sep = ""))
  colnames(transcriptomics) <- c('Gene','transcriptomics')
  
  temp1 <- merge(transcriptomics, phosphoproteomics, by = 'Gene', all = TRUE)
  temp2 <- merge(temp1, proteomics, by = 'Gene', all=TRUE)
  
  rownames(temp2) <- temp2$Gene
  
  temp2 <- temp2[, c("transcriptomics", "proteomics", "phosphoproteomics")]
  
  temp2[is.na(temp2)] <- 0
  
  write.table(unique(rownames(temp2)[temp2$transcriptomics>quantile(temp2$transcriptomics)[4]|temp2$phosphoproteomics>quantile(temp2$phosphoproteomics)[4]|temp2$proteomics>quantile(temp2$proteomics)[4]]), file = paste(contrast, "_relevant.csv"), row.names = FALSE, col.names = FALSE)
  
  write.table(unique(rownames(temp2)[temp2$transcriptomics>quantile(temp2$transcriptomics)[4]]), file = paste(contrast, "_transcriptomics_relevant.csv"), row.names = FALSE, col.names = FALSE)
  
  write.table(unique(rownames(temp2)[temp2$proteomics>quantile(temp2$proteomics)[4]]), file = paste(contrast, "_proteomics_relevant.csv"), row.names = FALSE, col.names = FALSE)
  
  write.table(unique(rownames(temp2)[temp2$phosphoproteomics>quantile(temp2$phosphoproteomics)[4]]), file = paste(contrast, "_phosphoproteomics_relevant.csv"), row.names = FALSE, col.names = FALSE)
}
makePaintOmics("1% O2, CyA-_vs_19% O2, CyA-.csv")
makePaintOmics("1% O2, CyA+_vs_1% O2, CyA-.csv")
makePaintOmics("19% O2, CyA-_vs_1% O2, CyA+.csv")

```
# Plot paintomics results


```{r}
#contrast1
contrast1 = read.csv(paste0(getwd(), "/contrast1/Paintomics_pathways_1O2NoCyA_vs_1O2PlusCyA.xls.csv"))
contrast1$Contrast = "X1O2NoCyA - X1O2PlusCyA"

#contrast2
contrast2 = read.csv(paste0(getwd(), "/contrast2/Paintomics_pathways_1O2NoCyclo_vs_19O2NoCyclo.xls.csv"))
contrast2$Contrast = "X1O2NoCyA - X19O2NoCyA"

#contrast3
contrast3 = read.csv(paste0(getwd(), "/contrast3/Paintomics_pathways_1O2PlusCyclo_vs_19O2NoCyclo.xls.csv"))
contrast3$Contrast = "X1O2PlusCyA - X19O2NoCyA"

results <- rbind(contrast1, contrast2, contrast3)
colnames(results) = c("Pathway.name", "Unique.genes", "Uniquemetabol", "Phosphoproteomics", "Proteomics", "Transcriptomics", "Combined", "X", "Contrast")
results = results[,!colnames(results) %in% c("X","Uniquemetabol")]
results = as.data.frame(apply(results, c(1,2), function(x){if(x=="-"){return(NA)}else{return(x)}}))
results$Proteomics_FDR = p.adjust(results$Proteomics, "fdr")
results$Phosphoproteomics_FDR = p.adjust(results$Phosphoproteomics, "fdr")
results$Transcriptomics_FDR = p.adjust(results$Transcriptomics, "fdr")
results$Combined_FDR = p.adjust(results$Combined, "fdr")
write.csv(results, "paintomics_allResults.csv")
head(results)
```

```{r}
melt_results = reshape2::melt(results, id.vars = c("Pathway.name", "Contrast"), measure.vars = c("Proteomics_FDR", "Phosphoproteomics_FDR", "Transcriptomics_FDR", "Combined_FDR"))
melt_results$variable = gsub("_FDR", "", melt_results$variable)
colnames(melt_results) = c("Pathway.name", "Contrast", "Dataset", "FDR")
melt_results$negativeLog10FDR = (-1)*log10(melt_results$FDR)
head(melt_results)
```

#old code, do not run

```{r, eval=FALSE}

results1 <- read.csv("Paintomics_pathways_1% O2, CyA-_vs_19% O2, CyA-.csv")
results1$Contrast <- "X19O2NoCyA - X1O2NoCyA"
results2 <- read.csv("Paintomics_pathways_1% O2, CyA+_vs_1% O2, CyA-.csv")
results2$Contrast <- "X1O2PlusCyA - X1O2NoCyA"
results3 <- read.csv("Paintomics_pathways_19% O2, CyA-_vs_1% O2, CyA+.csv")
results3$Contrast <- "X1O2NoCyA - X1O2PlusCyA"

results <- rbind(results1, results2, results3)

write.csv(results, "paintomics_allResults.csv")
head(results)

melt_results <- reshape2::melt(results, id.vars = c("Pathway.name", "Contrast"), measure.vars = c("Geneexpression.FDR.BH.", "Proteomics.FDR.BH.", "Proteomics.FDR.BH..1", "Combined.pValue.Fisher..FDR.BH."))
colnames(melt_results) <- c("Pathway", "Contrast", "Method", "FDR")
melt_results$negativelog10FDR <- -1*log10(melt_results$FDR)
levels(melt_results$Method) <- c("Transcriptomics", "Proteomics", "Phosphoproteomics", "Combined")
head(melt_results)
require(ggplot2)
temp <- melt_results$FDR < 0.05
melt_results <- na.omit(melt_results[temp,])
melt_results <- melt_results[order(melt_results$negativelog10FDR),]
ggplot(melt_results) + geom_point(aes(x=negativelog10FDR, y=Pathway, color=Contrast, shape=Contrast), size=3) + facet_wrap(facets = 'Method', nrow = 1) + theme_bw() + scale_color_brewer(palette = 'Set2')
ggsave(filename = "paintomics_allResults.svg", width = 12, height = 10, units = 'in')
ggsave(filename = "paintomics_allResults.png", width = 12, height = 10, units = 'in')
```



```{r}
require(ggplot2)
temp <- melt_results$FDR < 0.05
melt_results <- na.omit(melt_results[temp,])
melt_results <- melt_results[order(melt_results$negativeLog10FDR),]
ggplot(melt_results) + geom_point(aes(x=negativeLog10FDR, y=Pathway.name, color=Contrast, shape=Contrast), size=3) + facet_wrap(facets = 'Dataset', nrow = 1) + theme_bw() + scale_color_brewer(palette = 'Set2') 
ggsave(filename = "paintomics_allResults.svg", width = 12, height = 10, units = 'in')
ggsave(filename = "paintomics_allResults.png", width = 12, height = 10, units = 'in')
```



